### PR TITLE
linuxPackages.ena: 2.8.9 -> 2.12.0

### PR DESCRIPTION
**Pull request by kiike**
## Description of changes

This PR upgrades the Amazon Elastic Network Adapter driver to the latest version, also resolving the [broken build](https://hydra.nixos.org/build/259097959/nixlog/1).

ZHF: #309482

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by wegank** [comment]
Looks like `linuxPackages.ena` is failing?

**Comment by kiike** [comment]
> I think `linuxPackages` defaults to 6.6.

Ah I see what you mean, sorry. Yep, it fails building on 6.6.30:
<details>
<summary><code> nix build -L .#linuxPackages.ena
ena> Running phase: unpackPhase
ena> unpacking source archive /nix/store/7nnir1y4wikh13mrhrpdz0zk6ap9m4rn-source
ena> source root is source
ena> Running phase: patchPhase
ena> Running phase: updateAutotoolsGnuConfigScriptsPhase
ena> Running phase: configurePhase
ena> substituteStream(): WARNING: '--replace' is deprecated, use --replace-{fail,warn,quiet}. (file 'Makefile')
ena> Running phase: buildPhase
ena> build flags: SHELL=/nix/store/h3bhzvz9ipglcybbcvkxvm4vg9lwvqg4-bash-5.2p26/bin/bash O=\$\(buildRoot\) CC=/nix/store/md6hh4rrcrf99nssvcam3qaqs3skj086-gcc-wrapper-13.2.0/bin/cc HOSTCC=/nix/store/md6hh4rrcrf99nssvcam3qaqs3skj086-gcc-wrapper-13.2.0/bin/cc HOSTLD=/nix/store/m3gq2f6iwh3g67zrs27m7w82b37l8gf8-binutils-wrapper-2.41/bin/ld ARCH=x86_64
ena> ./configure.sh --kernel-dir /nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/build --verbose
ena> Trying to intercept CC arguments
ena> Detected compiler:  /nix/store/md6hh4rrcrf99nssvcam3qaqs3skj086-gcc-wrapper-13.2.0/bin/cc
ena> Managed to intercept CC arguments
ena> Command: /nix/store/md6hh4rrcrf99nssvcam3qaqs3skj086-gcc-wrapper-13.2.0/bin/cc -nostdinc -I/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/arch/x86/include -I./arch/x86/include/generated -I/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/include -I./include -I/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/arch/x86/include/uapi -I./arch/x86/include/generated/uapi -I/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/include/uapi -I./include/generated/uapi -include /nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/include/linux/compiler-version.h -include /nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/include/linux/kconfig.h -include /nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/include/linux/compiler_types.h -D__KERNEL__ -fmacro-prefix-map=/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/= -std=gnu11 -fshort-wchar -funsigned-char -fno-common -fno-PIE -fno-strict-aliasing -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx -fcf-protection=branch -fno-jump-tables -m64 -falign-jumps=1 -falign-loops=1 -mno-80387 -mno-fp-ret-in-387 -mpreferred-stack-boundary=3 -mskip-rax-setup -mtune=generic -mno-red-zone -mcmodel=kernel -fno-asynchronous-unwind-tables -mindirect-branch=thunk-extern -mindirect-branch-register -mindirect-branch-cs-prefix -mfunction-return=thunk-extern -fno-jump-tables -mharden-sls=all -fpatchable-function-entry=16,16 -fno-delete-null-pointer-checks -O2 -fno-allow-store-data-races -fstack-protector-strong -ftrivial-auto-var-init=zero -fno-stack-clash-protection -pg -mrecord-mcount -mfentry -DCC_USING_FENTRY -falign-functions=16 -fstrict-flex-arrays=3 -fno-strict-overflow -fno-stack-check -fconserve-stack -g -DMODULE -DKBUILD_BASENAME="main" -DKBUILD_MODNAME="main" -D__KBUILD_MODNAME=kmod_main -fno-pie -Werror=incompatible-pointer-types -Wno-nonnull -Werror=implicit -isystem /nix/store/9hgsinpfgyvsd92v0wlvmxv9wnaal68r-gcc-13.2.0/lib/gcc/x86_64-unknown-linux-gnu/13.2.0/include
ena> ### Detection of the available kernel API ###
ena> ENA_HAVE_PCI_DEV_ID: YES
ena> ENA_HAVE_XDP_DO_FLUSH: YES
ena> ENA_HAVE_CPUMASK_LOCAL_SPREAD: YES
ena> ENA_HAVE_UPDATE_AFFINITY_HINT: YES
ena> ENA_HAVE_ETHTOOL_PUTS: YES
ena> ENA_HAVE_ETHTOOL_RXFH_PARAM: NO
ena> ### Detection completed ###
ena> make -C /nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/build M=/build/source/kernel/linux/ena modules
ena> make[1]: Entering directory '/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/build'
ena>   CC [M]  /build/source/kernel/linux/ena/ena_netdev.o
ena>   CC [M]  /build/source/kernel/linux/ena/ena_ethtool.o
ena> /build/source/kernel/linux/ena/ena_ethtool.c: In function 'ena_metrics_stats_strings':
ena> /build/source/kernel/linux/ena/ena_ethtool.c:439:33: warning: implicit declaration of function 'ethtool_puts'; did you mean 'ethtool_notify'? [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wimplicit-function-declaration-Wimplicit-function-declaration8;;]
ena>   439 |                                 ethtool_puts(data, ena_metrics->name);
ena>       |                                 ^~~~~~~~~~~~
ena>       |                                 ethtool_notify
ena>   CC [M]  /build/source/kernel/linux/ena/ena_lpc.o
ena>   CC [M]  /build/source/kernel/linux/ena/ena_phc.o
ena>   CC [M]  /build/source/kernel/linux/ena/ena_xdp.o
ena>   CC [M]  /build/source/kernel/linux/ena/dim.o
ena>   CC [M]  /build/source/kernel/linux/ena/net_dim.o
ena>   CC [M]  /build/source/kernel/linux/ena/../common/ena_com//ena_com.o
ena>   CC [M]  /build/source/kernel/linux/ena/../common/ena_com//ena_eth_com.o
ena>   CC [M]  /build/source/kernel/linux/ena/ena_sysfs.o
ena>   LD [M]  /build/source/kernel/linux/ena/ena.o
ena>   MODPOST /build/source/kernel/linux/ena/Module.symvers
ena> ERROR: modpost: "ethtool_puts" [/build/source/kernel/linux/ena/ena.ko] undefined!
ena> make[3]: *** [/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/scripts/Makefile.modpost:145: /build/source/kernel/linux/ena/Module.symvers] Error 1
ena> make[2]: *** [/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/Makefile:1865: modpost] Error 2
ena> make[1]: *** [/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/Makefile:234: __sub-make] Error 2
ena> make[1]: Leaving directory '/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/build'
ena> make: *** [Makefile:102: ena.ko] Error 2
error: builder for '/nix/store/blxa9qml79w7lrm0h6d2vw07l57w39l9-ena-2.12.0-6.6.30.drv' failed with exit code 2;
       last 10 log lines:
       >   CC [M]  /build/source/kernel/linux/ena/../common/ena_com//ena_eth_com.o
       >   CC [M]  /build/source/kernel/linux/ena/ena_sysfs.o
       >   LD [M]  /build/source/kernel/linux/ena/ena.o
       >   MODPOST /build/source/kernel/linux/ena/Module.symvers
       > ERROR: modpost: "ethtool_puts" [/build/source/kernel/linux/ena/ena.ko] undefined!
       > make[3]: *** [/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/scripts/Makefile.modpost:145: /build/source/kernel/linux/ena/Module.symvers] Error 1
       > make[2]: *** [/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/Makefile:1865: modpost] Error 2
       > make[1]: *** [/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/Makefile:234: __sub-make] Error 2
       > make[1]: Leaving directory '/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/build'
       > make: *** [Makefile:102: ena.ko] Error 2
       For full logs, run 'nix log /nix/store/blxa9qml79w7lrm0h6d2vw07l57w39l9-ena-2.12.0-6.6.30.drv'.</code></summary></details>

I'll check if there's any workaround.

**Comment by kiike** [comment]
I am trying to trace the issue and from what I understand (not an expert), [the check here](https://github.com/amzn/amzn-drivers/blob/c77d43c7b6cfa1572b9ced44918d660d94252e90/kernel/linux/ena/config/test_defs.sh#L47) should take care of renaming the function at preprocessing time via [this definition here](https://github.com/amzn/amzn-drivers/blob/c77d43c7b6cfa1572b9ced44918d660d94252e90/kernel/linux/ena/kcompat.h#L1158). Unfortunately, during build, as you can see in the build above, the test to check if the kernel has the `ethtool_puts` API function succeeds when it shouldn't have. Makes sense?

**Comment by kiike** [comment]
I'm pushing a quick patch that overrides the setting from the `test_defs.sh` run and uses the kernel version to determine whether to patch the `ethtool_puts` function. Compilations for 6.6, 6.8 and 6.9 succeed now.

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 310680` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>9 packages failed to build:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.ena</li>
    <li>linuxPackages_4_19_hardened.ena (linuxKernel.packages.linux_4_19_hardened.ena)</li>
    <li>linuxKernel.packages.linux_5_10.ena</li>
    <li>linuxPackages_5_10_hardened.ena (linuxKernel.packages.linux_5_10_hardened.ena)</li>
    <li>linuxKernel.packages.linux_5_15.ena</li>
    <li>linuxPackages_5_15_hardened.ena (linuxKernel.packages.linux_5_15_hardened.ena)</li>
    <li>linuxKernel.packages.linux_5_4.ena</li>
    <li>linuxPackages_5_4_hardened.ena (linuxKernel.packages.linux_5_4_hardened.ena)</li>
    <li>linuxPackages_lqx.ena (linuxKernel.packages.linux_lqx.ena)</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_6_1.ena</li>
    <li>linuxPackages_6_1_hardened.ena (linuxKernel.packages.linux_6_1_hardened.ena)</li>
    <li>linuxPackages.ena (linuxKernel.packages.linux_6_6.ena)</li>
    <li>linuxPackages_hardened.ena (linuxPackages_6_6_hardened.ena)</li>
    <li>linuxKernel.packages.linux_6_8.ena</li>
    <li>linuxPackages_latest.ena (linuxKernel.packages.linux_6_9.ena)</li>
    <li>linuxPackages_latest-libre.ena (linuxKernel.packages.linux_latest_libre.ena)</li>
    <li>linuxPackages-libre.ena (linuxKernel.packages.linux_libre.ena)</li>
    <li>linuxPackages_xanmod.ena (linuxKernel.packages.linux_xanmod.ena)</li>
    <li>linuxPackages_xanmod_latest.ena (linuxKernel.packages.linux_xanmod_latest.ena ,linuxPackages_xanmod_stable.ena)</li>
    <li>linuxPackages_zen.ena (linuxKernel.packages.linux_zen.ena)</li>
  </ul>
</details>

**Comment by wegank** [comment]
There are still regressions: https://hydra.nixos.org/job/nixpkgs/trunk/linuxKernel.packages.linux_5_15.ena.x86_64-linux

**Comment by kiike** [comment]
Gave expanding the patch a go following the same pattern, but for older kernel versions. I can't use nixpkgs-review due to low memory, but `nix build` with the following vanilla versions work:

- 4.19
- 5.10
- 5.15
- 6.6
- 6.8
- 6.9

File: pkgs/os-specific/linux/ena/fix-ethtool-api.patch
Comment (wegank) [inline pkgs/os-specific/linux/ena/fix-ethtool-api.patch:1]: So I guess this file should be dropped?

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 310680` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>linuxPackages_lqx.ena (linuxKernel.packages.linux_lqx.ena)</li>
  </ul>
</details>
<details>
  <summary>19 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.ena</li>
    <li>linuxPackages_4_19_hardened.ena (linuxKernel.packages.linux_4_19_hardened.ena)</li>
    <li>linuxKernel.packages.linux_5_10.ena</li>
    <li>linuxPackages_5_10_hardened.ena (linuxKernel.packages.linux_5_10_hardened.ena)</li>
    <li>linuxKernel.packages.linux_5_15.ena</li>
    <li>linuxPackages_5_15_hardened.ena (linuxKernel.packages.linux_5_15_hardened.ena)</li>
    <li>linuxKernel.packages.linux_5_4.ena</li>
    <li>linuxPackages_5_4_hardened.ena (linuxKernel.packages.linux_5_4_hardened.ena)</li>
    <li>linuxKernel.packages.linux_6_1.ena</li>
    <li>linuxPackages_6_1_hardened.ena (linuxKernel.packages.linux_6_1_hardened.ena)</li>
    <li>linuxPackages.ena (linuxKernel.packages.linux_6_6.ena)</li>
    <li>linuxPackages_hardened.ena (linuxPackages_6_6_hardened.ena)</li>
    <li>linuxKernel.packages.linux_6_8.ena</li>
    <li>linuxPackages_latest.ena (linuxKernel.packages.linux_6_9.ena)</li>
    <li>linuxPackages_latest-libre.ena (linuxKernel.packages.linux_latest_libre.ena)</li>
    <li>linuxPackages-libre.ena (linuxKernel.packages.linux_libre.ena)</li>
    <li>linuxPackages_xanmod.ena (linuxKernel.packages.linux_xanmod.ena)</li>
    <li>linuxPackages_xanmod_latest.ena (linuxKernel.packages.linux_xanmod_latest.ena ,linuxPackages_xanmod_stable.ena)</li>
    <li>linuxPackages_zen.ena (linuxKernel.packages.linux_zen.ena)</li>
  </ul>
</details>

File: pkgs/os-specific/linux/ena/fix-ethtool-api.patch
Comment (kiike) [reply pkgs/os-specific/linux/ena/fix-ethtool-api.patch:1]: Correct. I am removing it now.

**Review by wegank — APPROVED** [review]

**Pull request merged**